content/en/blog/_posts/2023-03-08-memory-qos-cgroups-v2/index.md

```diff
@@ -0,0 +1,275 @@
+
+
+![](./container-memory-max.svg)  
+<sub>i: the i<sup>th</sup> container in one pod</sub>
+
+`memory.min` is mapped to memory requests, which results in reservation of memory resources
+that can never reclaimed by the kernel. This is how Memory QoS ensures the availability of 
+memory for Kubernetes pods. If there's no unprotected reclaimable memory available, OOM 
+killer is invoked.
+
+![](./container-memory-min.svg)  
+<sub>i: the i<sup>th</sup> container in one pod</sub>
+
+For memory protection, in addition to the original way of limiting memory usage, Memory QoS
+throttles workload approaching its memory limit, ensuring that the system is not overwhelmed
+by sporadic increases in memory usage. A new field, MemoryThrottlingFactor, is introduced in
+the KubeletConfiguration which is set to 0.9 by default. `memory.high` is mapped to throttling
+limit calculated by using MemoryThrottlingFactor, `spec.containers[].resources.requests.memory`
+and `spec.containers[].resources.limits.memory` as in the formula below, and rounding down the
+value to the nearest page size:
+
+![](./container-memory-high.svg)  
+<sub>i: the i<sup>th</sup> container in one pod</sub>
+
+**Note**: If a container has no memory limit specified, we substitute limits.memory for node allocatable memory.
+
+**Summary:**
+<table>
+    <tr>
+        <th style="text-align:center">File</th>
+        <th style="text-align:center">Description</th>
+   </tr>
+   <tr>
+        <td>memory.max</td>
+        <td><code>memory.max</code> specifies the maximum memory limit, a container is allowed to use. If a container tries to consume more memory than the configured limit, kernel terminates the process with an Out of Memory (OOM) error. 
+        <br>
+        <br>
+        <i>We map it to the container's memory limit</i>
+        </td>
+   </tr>
+   <tr>
+        <td>memory.min</td>
+        <td><code>memory.min</code> specifies a minimum amount of memory the cgroup must always retain, i.e., memory that can never be reclaimed by the system.
+        If there's no unprotected reclaimable memory available, OOM kill is invoked.
+        <br>
+        <br>
+        <i>We map it to the container's memory request</i>
+        </td>
+   </tr>
+   <tr>
+       <td>memory.high</td>
+       <td><code>memory.high</code> is the memory usage throttle limit. This is the main mechanism to control a cgroup's memory use. If a cgroup's memory use goes over the high boundary specified here, the cgroup’s processes are throttled and put under heavy reclaim pressure.
+       <br>
+       <br>
+       <i>We use a formula to calculate <code>memory.high</code>, depending on container's memory request, memory limit or node allocatable memory (if container's memory limit is empty) and a throttling factor. Please refer to the KEP for more details on the formula.</i>
+       </td>
+   </tr>
+</table>
+
+**Note** `memory.high` is set only on container level cgroups while memory.min is set on
```

Comment:
Please do consistently use code style when referring to a field name or property name.

```suggestion
**Note** `memory.high` is set only on container level cgroups while `memory.min` is set on
```

Reply:
Noted, will be careful from next time onwards.